pkg/relay/ethereum/upgrade.go

```diff
@@ -0,0 +1,97 @@
+package ethereum
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/datachainlab/ethereum-ibc-relay-chain/pkg/contract/iibcchannelupgradablemodule"
+	gethtypes "github.com/ethereum/go-ethereum/core/types"
+)
+
+func (c *Chain) ProposeUpgrade(
+	ctx context.Context,
+	portID string,
+	channelID string,
+	upgradeFields iibcchannelupgradablemodule.UpgradeFieldsData,
+	timeout iibcchannelupgradablemodule.TimeoutData,
+) error {
+	appAddr, err := c.ibcHandler.GetIBCModuleByChannel(c.CallOpts(ctx, 0), portID, channelID)
+	if err != nil {
+		return err
+	}
+
+	mockApp, err := iibcchannelupgradablemodule.NewIibcchannelupgradablemodule(
+		appAddr,
+		c.client.Client,
+	)
+	if err != nil {
+		return nil
+	}
+
+	txOpts, err := c.TxOpts(ctx, true)
+	if err != nil {
+		return err
+	}
+
+	tx, err := mockApp.ProposeUpgrade(
+		txOpts,
+		portID,
+		channelID,
+		upgradeFields,
+		timeout,
+	)
+	if err != nil {
+		return err
+	}
+
+	if receipt, err := c.client.WaitForReceiptAndGet(ctx, tx.Hash()); err != nil {
+		return err
+	} else if receipt.Status == gethtypes.ReceiptStatusFailed {
+		return fmt.Errorf("tx execution reverted")
```

Comment:
We should try to retrieve a revert reason from receipt here

Reply:
Sure. I will try to do that.

Reply:
@bluele done: https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/50/commits/2c0371da76e0fc01df339f5c78c0d79100b116d7